src/core/Akka.Streams/Dsl/Retry.cs

```diff
@@ -0,0 +1,370 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="Retry.cs" company="Akka.NET Project">
+//     Copyright (C) 2015-2017 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2017 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Akka.Annotations;
+using Akka.Pattern;
+using Akka.Streams.Stage;
+using Akka.Util;
+
+namespace Akka.Streams.Dsl
+{
+    public static class Retry
+    {
+        /// <summary>
+        /// EXPERIMENTAL API
+        /// <para>
+        /// Retry flow factory. given a flow that produces <see cref="Result{T}"/>s, this wrapping flow may be used to try
+        /// and pass failed elements through the flow again. More accurately, the given flow consumes a tuple
+        /// of `input` and `state`, and produces a tuple of <see cref="Result{T}"/> of `output` and `state`.
+        /// If the flow emits a failed element (i.e. <see cref="Result{T}.IsSuccess"/> is false), the <paramref name="retryWith"/>
+        /// function is fed with the `state` of the failed element, and may produce a new input-state tuple to pass through
+        /// the original flow. The function may also yield `null` instead of `(input, state)`, which means not to retry a failed element.
+        /// </para>
+        /// <para>
+        /// IMPORTANT CAVEAT:
+        /// The given flow must not change the number of elements passing through it (i.e. it should output
+        /// exactly one element for every received element). Ignoring this, will have an unpredicted result,
+        /// and may result in a deadlock.
+        /// </para> 
+        /// </summary>
+        /// <param name="flow">the flow to retry</param>
+        /// <param name="retryWith">if output was failure, we can optionaly recover from it,
+        /// and retry with a new pair of input and new state we get from this function.</param>
+        /// <typeparam name="TI">input elements type</typeparam>
+        /// <typeparam name="TS">state to create a new `(I,S)` to retry with</typeparam>
+        /// <typeparam name="TO">output elements type</typeparam>
+        /// <typeparam name="TM">materialized value type</typeparam>
+        [ApiMayChange]
+        public static IGraph<FlowShape<Tuple<TI, TS>, Tuple<Result<TO>, TS>>, TM> Create<TI, TS, TO, TM>(
```

Comment:
I know that his signature is quite long, but maybe we should stick to convention for generic type parameters (TIn, TState, TOut, TMat)?